DrawerKit/DrawerKit/Internal API/PresentationController.swift

```diff
@@ -59,106 +78,206 @@ private extension PresentationController {
     }
 
     var drawerPartialY: CGFloat {
         return containerViewH - drawerPartialH
     }
 
     var upperMarkY: CGFloat {
-        return drawerPartialY - upperMarkGap
+        return max(drawerPartialY - upperMarkGap, 0)
     }
 
     var lowerMarkY: CGFloat {
-        return drawerPartialY + lowerMarkGap
+        return min(drawerPartialY + lowerMarkGap, containerViewH)
+    }
+
+    var currentDrawerState: DrawerState {
+        get { return drawerState(for: currentDrawerY) }
+        set { currentDrawerY = drawerPositionY(for: newValue) }
+    }
+
+    func drawerPositionY(for state: DrawerState) -> CGFloat {
+        switch state {
+        case .collapsed:
+            return containerViewH
+        case .partiallyExpanded:
+            return drawerPartialY
+        case .fullyExpanded:
+            return 0
+        case let .transitioning(positionY):
+            return positionY
+        }
+    }
+
+    func drawerState(for positionY: CGFloat, clampToNearest: Bool = false) -> DrawerState {
+        if smallerThanOrEqual(positionY, 0) {
+            return .fullyExpanded
+        } else if greaterThanOrEqual(positionY, containerViewH) {
+            return .collapsed
+        } else if equal(positionY, drawerPartialY) {
+            return .partiallyExpanded
+        } else {
+            let posY = min(max(positionY, 0), containerViewH)
+            guard clampToNearest else { return .transitioning(posY) }
+            return drawerState(for: clamped(posY))
+        }
     }
 
     var currentDrawerY: CGFloat {
-        get { return presentedView?.frame.origin.y ?? 0 }
-        set { presentedView?.frame.origin.y = newValue }
+        get {
+            let posY = presentedView?.frame.origin.y ?? 0
+            return min(max(posY, 0), containerViewH)
+        }
+        set {
+            let posY = min(max(newValue, 0), containerViewH)
+            presentedView?.frame.origin.y = posY
+        }
     }
 
     var currentDrawerCornerRadius: CGFloat {
-        get { return presentedView?.layer.cornerRadius ?? 0 }
+        get {
+            let radius = presentedView?.layer.cornerRadius ?? 0
+            return min(max(radius, 0), maximumCornerRadius)
+        }
         set {
-            presentedView?.layer.cornerRadius = newValue
+            let radius = min(max(newValue, 0), maximumCornerRadius)
+            presentedView?.layer.cornerRadius = radius
             if #available(iOS 11.0, *) {
                 presentedView?.layer.maskedCorners =
                     [.layerMinXMinYCorner, .layerMaxXMinYCorner]
             } else {
-                presentedView?.roundCorners([.topLeft, .topRight], radius: newValue)
+                presentedView?.roundCorners([.topLeft, .topRight], radius: radius)
+            }
+        }
+    }
+
+    func cornerRadius(at state: DrawerState) -> CGFloat {
+        guard maximumCornerRadius != 0 && drawerPartialY != 0
+            && drawerPartialY != containerViewH else { return 0 }
```

Comment:
```
guard maximumCornerRadius != 0
    && drawerPartialY != 0
    && drawerPartialY != containerViewH
else { return 0 }
```

Reply:
Fixed in https://github.com/Babylonpartners/DrawerKit/pull/27/commits/4d921097164ae06ad46d0b2a993339f0f0358300.